api/src/main/java/org/apache/iceberg/catalog/SupportsCatalogTransactions.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.catalog;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Catalog methods for working with catalog-level transactions.
+ *
+ * <p>Catalog implementations are not required to support catalog-level transactional state.
+ * If they do, they may support one or more {@code IsolationLevel}s and one or more
+ * {@code LockingMode}s.
+ */
+public interface SupportsCatalogTransactions {
+
+  /**
+   * The level of isolation for a transaction.
+   */
+  enum IsolationLevel {
+
+    /**
+     * Only read committed data. Reading the same table multiple times
+     * may result in different committed reads for each read.
+     */
+    READ_COMMITTED,
+
+    /**
+     * Only read committed data. Reading the same table multiple times will
+     * result in the same view of those tables.
+     */
+    REPEATED_READ,
+
+    /**
+     * Only read committed data. A commit will only succeed if there have
+     * been no changes to data read during the course of the transaction
+     * prior to the commit operation.
+     */
+    SERIALIZABLE;
+  }
+
+  /**
+   * The type of locking mode used by the transaction.
+   */
+  enum LockingMode {
+
+    /**
+     * Pessimistically lock tables. In this situation, each table being worked on will grab a lock.
+     * This mode will typically result in a better chance for transactions to complete at the cost
+     * of throughput and resource consumption.
+     */
+    PESSIMISTIC,
+
+    /**
+     * Run the transaction with an optimistic behavior. This assumes that most operations will not
+     * conflict. It typically increases concurrency and throughput while reducing resource
+     * consumption.
+     */
+    OPTIMISTIC;
+  }
+
+  /**
+   * Get the list of {@link LockingMode}s this Catalog supports.
+   * @return A set of locking modes supported.
+   */
+  default Set<LockingMode> lockingModes() {
+    return Collections.emptySet();
+  }
+
+  /**
+   * Get the list of {@link IsolationLevel}s this Catalog supports.
+   * @return A set of locking modes supported.
+   */
+  default Set<IsolationLevel> isolationLevels() {
+    return Collections.emptySet();
```

Comment:
I'm not sure about the defaults here. If no isolation levels are supported, then what is the point? I think that once we know more about the implementation, we'll be able to say what isolation level is the lowest that can be supported, and use that for the default.

Locking mode is similar, but I would assume that it can be defaulted to a set with just `OPTIMISTIC`.

Reply:
Yeah, agreed. I was thinking that after I wrote them :D